File: modules/word-widget/ios/WordWidgetModule.swift
import ExpoModulesCore
import WidgetKit

/// The app's side of the iOS home-screen widget.
///
/// The widget runs in its own process and cannot read the app's storage, so
/// everything it shows is pushed here into a shared App Group. Keep the keys
/// in step with targets/widget/WordDeck.swift.
public class WordWidgetModule: Module {
  private static let suiteName = "group.app.kalima.vocab"

  public func definition() -> ModuleDefinition {
    Name("WordWidget")

    Function("isSupported") {
      // WidgetKit exists from iOS 14; the target itself asks for 16.
      if #available(iOS 16.0, *) { return true } else { return false }
    }

    /**
     * iOS can only answer this asynchronously, and the caller wants it
     * synchronously to decide what to render. Zero means "don't claim
     * anything" — the settings screen then shows how to add the widget rather
     * than asserting one is already placed.
     */
    Function("placedCount") { 0 }

    AsyncFunction("setDeck") { (cards: [[String: String]]) in
      let payload = cards.map { card in
        [
          "w": card["word"] ?? "",
          "t": card["translation"] ?? "",
          "p": card["ipa"] ?? "",
          "i": card["id"] ?? "",
        ]
      }

      guard
        let defaults = UserDefaults(suiteName: Self.suiteName),
        let data = try? JSONSerialization.data(withJSONObject: payload),
        let json = String(data: data, encoding: .utf8)
      else { return }

      defaults.set(json, forKey: "deck")
      // a fresh deck starts from its first card
      defaults.set(0, forKey: "base")

      if #available(iOS 14.0, *) {
        WidgetCenter.shared.reloadAllTimelines()
      }
    }

    AsyncFunction("refresh") {
      if #available(iOS 14.0, *) {
        WidgetCenter.shared.reloadAllTimelines()
      }
    }
  }
}

File: modules/word-widget/android/src/main/java/expo/modules/wordwidget/WordWidgetModule.kt
package expo.modules.wordwidget

import android.appwidget.AppWidgetManager
import android.content.ComponentName
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

class WidgetCard : Record {
  @Field val word: String = ""
  @Field val translation: String = ""
  @Field val ipa: String = ""
  @Field val id: String = ""
}

/**
 * The app's side of the home-screen widget.
 *
 * Android only. iOS widgets are a separate WidgetKit extension and are not
 * part of this module — the JS wrapper reports `isSupported() = false` there
 * so the feature hides itself rather than half-working.
 */
class WordWidgetModule : Module() {

  private val context
    get() = requireNotNull(appContext.reactContext)

  override fun definition() = ModuleDefinition {
    Name("WordWidget")

    Function("isSupported") { true }

    /** How many widgets the user has actually placed. */
    Function("placedCount") {
      AppWidgetManager.getInstance(context).getAppWidgetIds(
        ComponentName(context, WordWidgetProvider::class.java),
      ).size
    }

    AsyncFunction("setDeck") { cards: List<WidgetCard> ->
      WidgetStore.saveDeck(
        context,
        cards.map {
          WidgetWord(
            word = it.word,
            translation = it.translation,
            ipa = it.ipa,
            id = it.id,
          )
        },
      )
      WordWidgetProvider.renderAll(context)
    }

    AsyncFunction("refresh") {
      WordWidgetProvider.renderAll(context)
    }
  }
}
